walkeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,148 @@
+
+use std::{path::PathBuf};
+use std::{thread, time};
+
+use tokio::runtime::{Builder};
+
+
+use anyhow::Result;
+
+use tokio::sync::watch::Receiver;
+use zenith_utils::lsn::Lsn;
+use tracing::*;
+
+
+/// 
+/// High level abstraction could be an async task service with tasks executed on a thread pool
+/// task service doesn't have any state (other than in-memory), requests to process are equeued 
+/// 
+/// Wal Backup will have 2 interfaces:
+/// 1. discover local state and segments in S3, enqueue each missing segment for upload
+/// 2. provide enqueue interface for on-demand requests
+/// 3. When queue is full we shuould introduce a spill interface
+/// 4. uploads should be in order of creation
+/// 5. service should have a knob for parallelism (as single threaded upload is slow)
+
+
+#[allow(dead_code)]
+pub struct Seg {
+    lsn: Lsn,
+    timeline: u32,
+    name: PathBuf,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct WalBackup {
+    
+    _remote_storage : u32,
+    // _async_service : Runtime,
+    // have a per-timeline status of what segments are uploaded
+    // Given segments have very specific naming we store it as a bitmap (we probably don't care about postgres timeline in this case, all we care is the start LSN)
+    // we check the S3 state only when we're a leader and the timeline is initialized.
+}
+
+#[allow(unreachable_code)]
+async fn detect_task(mut event: Receiver<Lsn>, backup: WalBackup) -> Result<()> {
+
+    while event.changed().await.is_ok() {
+        warn!("Woken Up for segment backup");
+        backup.backup_stuff(*event.borrow())?;
+    }
+
+    Ok(())
+}
+
+async fn upload_task(s: String, lsn: Lsn) -> Result<()> {
+    warn!("waiting for lsn {} ", lsn);
+    thread::sleep(time::Duration::from_millis(1000));
+
+    warn!("uploading {} ", s);
+    thread::sleep(time::Duration::from_millis(1000));
+    warn!("uploaded {}", s);
+    Ok(())
+}
+
+
+// #[allow(dead_code)]
+impl WalBackup {
+
+
+    pub fn create(event: Receiver<Lsn>) -> Self {
+        warn!("Backup service is created");
+
+        let runtime = Builder::new_multi_thread()
```

Comment:
I think the problem is that runtime is dropped at the end of this function. If runtime is dropped it cancels all the tasks it owns, which  leads to receiver of the channel being dropped

Reply:
I was not able to reproduce the exact error though. What did you do before it happened? I tried your `restart_all.sh` script, data insertion operations return errors from pageserver probably because wal didn't reach pageserver side